include/deal.II/base/tensor_accessors.h

```diff
@@ -0,0 +1,586 @@
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__tensor_accessors_h
+#define dealii__tensor_accessors_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/template_constraints.h>
+#include <deal.II/base/table_indices.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * This namespace is a collection of algorithms working on generic
+ * tensorial objects.
+ *
+ * A tensorial object has the notion of a rank and allows a rank-times
+ * recursive application of the index operator, e.g., if <code>t</code> is
+ * a tensorial object of rank 4, the following access is valid:
+ * @code
+ *   t[1][2][1][4]
+ * @endcode
+ *
+ * deal.II has its own implementation for tensorial objects such as
+ * dealii::Tensor<rank, dim, Number> and
+ * dealii::SymmetricTensor<rank, dim, Number>
+ *
+ * The methods and algorithms implemented in this namespace, however, are
+ * fully generic. More precisely, a minimal interface a tensorial object of
+ * type <code>T</code> must provide is a local typedef
+ * <code>value_type</code> and an index operator <code>operator[]()</code>
+ * that returns a (const or non-const) reference of
+ * <code>value_type</code>:
+ * @code
+ *   template<...>
+ *   class T
+ *   {
+ *     typedef ... value_type;
+ *     value_type & operator[](unsigned int);
+ *     const value_type & operator[](unsigned int) const;
+ *   };
+ * @endcode
+ *
+ * This namespace provides primitves for access, reordering and contraction
+ * of such objects.
+ *
+ * @ingroup LAC
+ * @ingroup geomprimitives
+ */
+namespace TensorAccessors
+{
+  /**
+   * This class provides a local typedef @p value_type that is equal to
+   * <code>T::value_type</code> if the template parameter @tparam T does not
+   * have a const qualifier.
+   *
+   * If @tparam T has a const qualifier a const is also added to the local
+   * typdef @p value_type, more precisely, it is defined as
+   * <code>const T::value_type</code>.
+   *
+   * This type wrapper is useful to automatically add a const qualifier to
+   * return values of reference types.
+   *
+   * @author Matthias Maier, 2015
+   */
+  template <typename T>
+  struct AccessType
+  {
+    typedef typename T::value_type value_type;
+  };
+
+
+  template <typename T>
+  struct AccessType<const T>
+  {
+    typedef const typename T::value_type value_type;
+  };
+
+
+  /**
+   * This class provides a local typedef @p value_type that is equal to the
+   * typedef <code>value_type</code> after @tparam rank - 1 recursive
+   * applications. Further, constness is preserved via the AccessType type
+   * trait, i.e., if T is const, ReturnType<rank, T>::value_type will also
+   * be const.
+   *
+   * @author Matthias Maier, 2015
+   */
+  template <int rank, typename T>
+  struct ReturnType
+  {
+    typedef typename ReturnType<rank - 1, typename AccessType<T>::value_type>::value_type value_type;
```

Comment:
I think you should name the first template parameter something different than rank, because it's neither the rank of the object you're indexing, nor the rank of the resulting object. Maybe `deref_steps` or something?